utilities/serviceutils.py

```diff
@@ -108,31 +107,28 @@ async def send(self, content):
         raise NotImplementedError()
 
 
+@dataclass
 class ServiceMessage:
-    def __init__(
-        self, id: str, content: str, author: ServiceUser, channel: ServiceChannel, service: Services
-    ):
-        self.content = content
-        self.author = author
-        self.channel = channel
-        self.service = service
-        self.clean_content = content
-        self.service = service
-        self.created_at = datetime.now(timezone.utc)
-        self.id = id
-        self.mentions: list[ServiceUser] = []
-        self.reference: Optional[ServiceMessage] = None
-        self.is_dm = False
-
+    id: str
+    content: str
+    author: ServiceUser
+    channel: ServiceChannel
+    service: Services
+    clean_content: str = field(default="", init=False)
+    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc), init=False)
+    mentions: list[ServiceUser] = field(default_factory=list, init=False)
+    reference: Optional["ServiceMessage"] = field(default=None, init=False)
```

Comment:
hmm, static analysis tools shouldn't need the [forward reference](https://peps.python.org/pep-0484/#forward-references) type hint syntax with quotes, do we use any tool that needs it?

if yes, have you checked if other places in the code need it too, please?

Reply:
If an attribute of a class has the same type as its class it needs to be quoted, otherwise it's going to throw an error. Same if you want to type annotate an argument or return type of one of its methods (e.g. `def __eq__(self, other: "C")`. You can omit this need by just adding `from __future__ import annotations` at the top of the file and I'm with that option too.

Reply:
ah, `@dataclass` itself will throw an error or which tool? type systems inside VS Code and IntelliJ both seemed fine without the quotes AFAICT 😅

Reply:
> it is one more layer of abstraction to imagine the autogenerated init from the decorator

I think defining what attributes a class should have and having `__init__` autogenerated from that increases code clarity. I don't think imagining what arguments should be passed to `__init__` is a problem. You just make sure to add all `field`s with `init=False` below those with (implicit) `init=True` and whenever you create an instance of this class or call `super()__init__`, your IDE displays it to you.

Reply:
> ah, `@dataclass` itself will throw an error or which tool? type systems inside VS Code and IntelliJ both seemed fine without the quotes AFAICT sweat_smile

`@dataclass` itself

![Screenshot from 2022-09-17 11-04-50](https://user-images.githubusercontent.com/57115520/190849224-7449180e-2937-4720-9b62-1d94c1c4803d.png)

Reply:
Oh how I long for `from __future__ import annotations` to just be the default.  Alternatively to quoting we actually could just import this.  Are there any drawbacks to doing so?